[Skia] Return early when receiving an all-zero DashArray

spenap opened the pull request:
#### 302e7f17d0ef26379d41ca0a5c79677bca01bfa4
<pre>
[Skia] Return early when receiving an all-zero DashArray
<a href="https://bugs.webkit.org/show_bug.cgi?id=285610">https://bugs.webkit.org/show_bug.cgi?id=285610</a>

Reviewed by Carlos Garcia Campos.

Validate dash array elements according to the spec, including
optimising the all-zero case. An all-zero DashArray can be detected and skipped.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::isValidDashArray): New helper that checks array is valid.
(WebCore::GraphicsContextSkia::setLineDash): Use the helper to assign a
nullptr and return early.

Canonical link: <a href="https://commits.webkit.org/289164@main">https://commits.webkit.org/289164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/443bb82a29711c92b9b9176c1db486b54234cdef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39944 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90637 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36551 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87597 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13223 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66481 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24296 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88557 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77664 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46764 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4038 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31927 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35621 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74719 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32778 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92297 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12864 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75197 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13084 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73498 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74333 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18357 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18569 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16999 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4965 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12882 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18269 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12686 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16121 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14463 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -906,9 +907,22 @@ void GraphicsContextSkia::setLineCap(LineCap lineCap)
+        total += dash;
+    }
+    // Nothing to be done for all-zero or empty arrays
+    return total > 0;
```
@carlosgcampos , @Scony 

I took the letter of the spec in https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-setlinedash, and felt OK about checking for negative and not-finite values in the loop. I then took a look at what Skia does when checking for `total > 0` as that covers empty and all-zero arrays.

While there's some overhead in doing this -- since the Skia internal implementation does it, too, it makes sense, as we're now enforcing what the spec dictates. What do you think?

spenap replied:
The test failures are in relevant areas, so I've marked this as a draft.

Scony replied:
I'd include all the checks here is better for readability.

Scony commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -906,9 +907,22 @@ void GraphicsContextSkia::setLineCap(LineCap lineCap)
 void GraphicsContextSkia::setLineDash(const DashArray& dashArray, float dashOffset)
 {
-    if (dashArray.isEmpty()) {
+    if (isValidDashArray(dashArray)) {
```
shouldn't it be `!isValidDashArray(dashArray)` ?

spenap replied:
Indeed it should! I started writing this as a `isDashArrayAllZeros` and as I updated it to cover all the criteria from the spec I forgot to change the check :sweat_smile:

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -906,9 +907,22 @@ void GraphicsContextSkia::setLineCap(LineCap lineCap)
 
+static bool isValidDashArray(const DashArray& dashArray)
+{
+    // See 'dom-context-2d-setlinedash': if the array contains not finite or negative values, return
```
```suggestion
    // See 'dom-context-2d-setlinedash': if the array contains not finite or negative values, return.
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -906,9 +907,22 @@ void GraphicsContextSkia::setLineCap(LineCap lineCap)
+            return false;
+        total += dash;
+    }
+    // Nothing to be done for all-zero or empty arrays
```
```suggestion
    // Nothing to be done for all-zero or empty arrays.
```

spenap committed:
[Skia] Return early when receiving an all-zero DashArray
https://bugs.webkit.org/show_bug.cgi?id=285610

Reviewed by Carlos Garcia Campos.

Validate dash array elements according to the spec, including
optimising the all-zero case. An all-zero DashArray can be detected and skipped.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::isValidDashArray): New helper that checks array is valid.
(WebCore::GraphicsContextSkia::setLineDash): Use the helper to assign a
nullptr and return early.

Canonical link: https://commits.webkit.org/289164@main
